File: mobile/ios/Tolkflip/Encryption/E2EEncryptionManager.swift
import Foundation
import SignalProtocol

/**
 * End-to-end encryption manager using Signal Protocol for iOS
 */
class E2EEncryptionManager {
    static let shared = E2EEncryptionManager()

    private var protocolStore: SignalProtocolStoreImpl?
    private var userId: String?

    private let preKeyStartId = 1
    private let preKeyCount = 100

    private init() {}

    /**
     * Initialize the encryption manager for a user
     */
    func initialize(userId: String) {
        self.userId = userId
        self.protocolStore = SignalProtocolStoreImpl(userId: userId)

        // Generate keys if needed
        if protocolStore?.getIdentityKeyPair() == nil {
            generateIdentityKeys()
        }

        // Generate pre-keys if needed
        if protocolStore?.loadSignedPreKey(signedPreKeyId: 1) == nil {
            generatePreKeys()
        }

        print("E2E encryption initialized for user: \(userId)")
    }

    /**
     * Generate identity key pair
     */
    private func generateIdentityKeys() {
        guard let store = protocolStore else { return }

        let identityKeyPair = IdentityKeyPair.generate()
        store.saveIdentityKeyPair(identityKeyPair)

        // Generate registration ID
        let registrationId = UInt32.random(in: 1...16380)
        store.saveRegistrationId(registrationId)

        print("Generated new identity keys")
    }

    /**
     * Generate pre-keys for key exchange
     */
    private func generatePreKeys() {
        guard let store = protocolStore else { return }
        guard let identityKeyPair = store.getIdentityKeyPair() else { return }

        // Generate signed pre-key
        let signedPreKeyPair = KeyPair.generate()
        let timestamp = Date().timeIntervalSince1970

        // Sign the public key
        let signature = try? identityKeyPair.privateKey.sign(message: signedPreKeyPair.publicKey.serialize())

        let signedPreKeyRecord = SignedPreKeyRecord(
            id: 1,
            timestamp: UInt64(timestamp * 1000),
            keyPair: signedPreKeyPair,
            signature: signature ?? Data()
        )

        store.storeSignedPreKey(signedPreKeyRecord, signedPreKeyId: 1)

        // Generate one-time pre-keys
        for i in preKeyStartId..<(preKeyStartId + preKeyCount) {
            let preKeyPair = KeyPair.generate()
            let preKeyRecord = PreKeyRecord(id: UInt32(i), keyPair: preKeyPair)
            store.storePreKey(preKeyRecord, preKeyId: UInt32(i))
        }

        print("Generated \(preKeyCount) pre-keys")
    }

    /**
     * Get pre-key bundle for sharing with other users
     */
    func getPreKeyBundle() -> PreKeyBundle? {
        guard let store = protocolStore else { return nil }
        guard let identityKeyPair = store.getIdentityKeyPair() else { return nil }
        guard let signedPreKey = store.loadSignedPreKey(signedPreKeyId: 1) else { return nil }

        let registrationId = store.getLocalRegistrationId()

        // Get a random one-time pre-key
        let preKeyId = UInt32(preKeyStartId + Int.random(in: 0..<preKeyCount))
        guard let preKey = store.loadPreKey(preKeyId: preKeyId) else { return nil }

        return PreKeyBundle(
            registrationId: registrationId,
            deviceId: 1,
            preKeyId: preKeyId,
            preKeyPublic: preKey.keyPair.publicKey,
            signedPreKeyId: 1,
            signedPreKeyPublic: signedPreKey.keyPair.publicKey,
            signedPreKeySignature: signedPreKey.signature,
            identityKey: identityKeyPair.publicKey
        )
    }

    /**
     * Process received pre-key bundle from another user
     */
    func processPreKeyBundle(recipientId: String, bundle: PreKeyBundle) {
        guard let store = protocolStore else { return }

        let address = SignalAddress(name: recipientId, deviceId: 1)
        let sessionBuilder = SessionBuilder(store: store, address: address)

        do {
            try sessionBuilder.process(preKeyBundle: bundle)
            print("Processed pre-key bundle for \(recipientId)")
        } catch {
            print("Error processing pre-key bundle: \(error)")
        }
    }

    /**
     * Encrypt a message for a recipient
     */
    func encryptMessage(recipientId: String, plaintext: String) -> String? {
        guard let store = protocolStore else { return nil }

        let address = SignalAddress(name: recipientId, deviceId: 1)
        let sessionCipher = SessionCipher(store: store, address: address)

        do {
            guard let plaintextData = plaintext.data(using: .utf8) else { return nil }
            let ciphertext = try sessionCipher.encrypt(plaintextData)

            // Convert to base64
            let encoded = ciphertext.serialize().base64EncodedString()

            // Prepend message type (3 = PreKeySignalMessage, 2 = SignalMessage)
            let prefix = ciphertext.type == .preKey ? "3:" : "2:"

            print("Encrypted message for \(recipientId)")
            return prefix + encoded

        } catch {
            print("Error encrypting message: \(error)")
            return nil
        }
    }

    /**
     * Decrypt a received message
     */
    func decryptMessage(senderId: String, ciphertextString: String) -> String? {
        guard let store = protocolStore else { return nil }

        // Parse message type and ciphertext
        let parts = ciphertextString.components(separatedBy: ":")
        guard parts.count == 2 else {
            print("Invalid ciphertext format")
            return nil
        }

        guard let messageType = Int(parts[0]),
              let ciphertextData = Data(base64Encoded: parts[1]) else {
            print("Invalid ciphertext data")
            return nil
        }

        let address = SignalAddress(name: senderId, deviceId: 1)
        let sessionCipher = SessionCipher(store: store, address: address)

        do {
            let plaintext: Data

            if messageType == 3 {
                // PreKey message
                let preKeyMessage = try PreKeySignalMessage(bytes: ciphertextData)
                plaintext = try sessionCipher.decrypt(preKeyMessage: preKeyMessage)
            } else {
                // Regular message
                let signalMessage = try SignalMessage(bytes: ciphertextData)
                plaintext = try sessionCipher.decrypt(signalMessage: signalMessage)
            }

            print("Decrypted message from \(senderId)")
            return String(data: plaintext, encoding: .utf8)

        } catch {
            print("Error decrypting message: \(error)")
            return nil
        }
    }

    /**
     * Check if we have a session with a user
     */
    func hasSession(userId: String) -> Bool {
        guard let store = protocolStore else { return false }
        let address = SignalAddress(name: userId, deviceId: 1)
        return store.containsSession(for: address)
    }

    /**
     * Delete session with a user
     */
    func deleteSession(userId: String) {
        guard let store = protocolStore else { return }
        let address = SignalAddress(name: userId, deviceId: 1)
        store.deleteSession(for: address)
        print("Deleted session with \(userId)")
    }

    /**
     * Get the public identity key as a base64 string
     */
    func getPublicIdentityKey() -> String? {
        guard let store = protocolStore else { return nil }
        guard let identityKeyPair = store.getIdentityKeyPair() else { return nil }
        return identityKeyPair.publicKey.serialize().base64EncodedString()
    }

    /**
     * Save another user's public identity key
     */
    func savePublicIdentityKey(userId: String, publicKeyBase64: String) {
        guard let store = protocolStore else { return }
        guard let publicKeyData = Data(base64Encoded: publicKeyBase64) else { return }

        do {
            let identityKey = try PublicKey(bytes: publicKeyData)
            let address = SignalAddress(name: userId, deviceId: 1)
            _ = store.saveIdentity(identityKey, for: address)
            print("Saved public identity key for \(userId)")
        } catch {
            print("Error saving public identity key: \(error)")
        }
    }

    /**
     * Clear all encryption data (for logout)
     */
    func clearAll() {
        protocolStore?.deleteAllSessions()
        print("Cleared all encryption data")
    }
}

// MARK: - Supporting Types

struct PreKeyBundle {
    let registrationId: UInt32
    let deviceId: UInt32
    let preKeyId: UInt32
    let preKeyPublic: PublicKey
    let signedPreKeyId: UInt32
    let signedPreKeyPublic: PublicKey
    let signedPreKeySignature: Data
    let identityKey: PublicKey
}

struct PreKeyRecord {
    let id: UInt32
    let keyPair: KeyPair

    func serialize() -> Data {
        // Serialize the record for storage
        var data = Data()
        data.append(withUnsafeBytes(of: id.bigEndian) { Data($0) })
        data.append(keyPair.publicKey.serialize())
        data.append(keyPair.privateKey.serialize())
        return data
    }

    static func deserialize(_ data: Data) throws -> PreKeyRecord {
        guard data.count >= 4 else { throw EncryptionError.invalidData }

        let id = data[0..<4].withUnsafeBytes { $0.load(as: UInt32.self).bigEndian }
        let publicKey = try PublicKey(bytes: data[4..<37])
        let privateKey = try PrivateKey(bytes: data[37...])

        return PreKeyRecord(id: id, keyPair: KeyPair(publicKey: publicKey, privateKey: privateKey))
    }
}

struct SignedPreKeyRecord {
    let id: UInt32
    let timestamp: UInt64
    let keyPair: KeyPair
    let signature: Data

    func serialize() -> Data {
        var data = Data()
        data.append(withUnsafeBytes(of: id.bigEndian) { Data($0) })
        data.append(withUnsafeBytes(of: timestamp.bigEndian) { Data($0) })
        data.append(keyPair.publicKey.serialize())
        data.append(keyPair.privateKey.serialize())
        data.append(signature)
        return data
    }

    static func deserialize(_ data: Data) throws -> SignedPreKeyRecord {
        guard data.count >= 12 else { throw EncryptionError.invalidData }

        let id = data[0..<4].withUnsafeBytes { $0.load(as: UInt32.self).bigEndian }
        let timestamp = data[4..<12].withUnsafeBytes { $0.load(as: UInt64.self).bigEndian }
        let publicKey = try PublicKey(bytes: data[12..<45])
        let privateKey = try PrivateKey(bytes: data[45..<77])
        let signature = data[77...]

        return SignedPreKeyRecord(
            id: id,
            timestamp: timestamp,
            keyPair: KeyPair(publicKey: publicKey, privateKey: privateKey),
            signature: signature
        )
    }
}

enum EncryptionError: Error {
    case invalidData
    case keyGenerationFailed
    case encryptionFailed
    case decryptionFailed
}

File: mobile/android/app/src/main/java/com/tolkflip/encryption/E2EEncryptionManager.java
package com.tolkflip.encryption;

import android.content.Context;
import android.util.Base64;
import android.util.Log;

import org.whispersystems.libsignal.DuplicateMessageException;
import org.whispersystems.libsignal.IdentityKey;
import org.whispersystems.libsignal.IdentityKeyPair;
import org.whispersystems.libsignal.InvalidKeyException;
import org.whispersystems.libsignal.InvalidKeyIdException;
import org.whispersystems.libsignal.InvalidMessageException;
import org.whispersystems.libsignal.InvalidVersionException;
import org.whispersystems.libsignal.LegacyMessageException;
import org.whispersystems.libsignal.NoSessionException;
import org.whispersystems.libsignal.SessionCipher;
import org.whispersystems.libsignal.SignalProtocolAddress;
import org.whispersystems.libsignal.UntrustedIdentityException;
import org.whispersystems.libsignal.ecc.Curve;
import org.whispersystems.libsignal.ecc.ECKeyPair;
import org.whispersystems.libsignal.protocol.CiphertextMessage;
import org.whispersystems.libsignal.protocol.PreKeySignalMessage;
import org.whispersystems.libsignal.protocol.SignalMessage;
import org.whispersystems.libsignal.state.PreKeyBundle;
import org.whispersystems.libsignal.state.PreKeyRecord;
import org.whispersystems.libsignal.state.SignalProtocolStore;
import org.whispersystems.libsignal.state.SignedPreKeyRecord;
import org.whispersystems.libsignal.util.KeyHelper;

import java.io.IOException;
import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.List;

/**
 * End-to-end encryption manager using Signal Protocol
 */
public class E2EEncryptionManager {
    private static final String TAG = "E2EEncryptionManager";
    private static final int PREKEY_START_ID = 1;
    private static final int PREKEY_COUNT = 100;

    private final Context context;
    private final SignalProtocolStore protocolStore;
    private final String userId;

    private static E2EEncryptionManager instance;

    private E2EEncryptionManager(Context context, String userId) {
        this.context = context.getApplicationContext();
        this.userId = userId;
        this.protocolStore = new SignalProtocolStoreImpl(context, userId);
    }

    public static synchronized E2EEncryptionManager getInstance(Context context, String userId) {
        if (instance == null || !instance.userId.equals(userId)) {
            instance = new E2EEncryptionManager(context, userId);
        }
        return instance;
    }

    /**
     * Initialize the encryption manager (generate keys if needed)
     */
    public void initialize() {
        try {
            // Check if identity key pair exists
            IdentityKeyPair identityKeyPair = protocolStore.getIdentityKeyPair();

            if (identityKeyPair == null) {
                Log.i(TAG, "Generating new identity key pair");
                identityKeyPair = KeyHelper.generateIdentityKeyPair();
                protocolStore.saveIdentity(new SignalProtocolAddress(userId, 1), identityKeyPair.getPublicKey());
            }

            // Generate pre-keys if needed
            if (protocolStore.loadSignedPreKey(1) == null) {
                generatePreKeys();
            }

            Log.i(TAG, "E2E encryption initialized");
        } catch (Exception e) {
            Log.e(TAG, "Error initializing encryption", e);
        }
    }

    /**
     * Generate pre-keys for key exchange
     */
    private void generatePreKeys() throws InvalidKeyException {
        IdentityKeyPair identityKeyPair = protocolStore.getIdentityKeyPair();

        // Generate signed pre-key
        int signedPreKeyId = 1;
        ECKeyPair signedPreKeyPair = Curve.generateKeyPair();
        long timestamp = System.currentTimeMillis();
        byte[] signedPreKeySignature = Curve.calculateSignature(
            identityKeyPair.getPrivateKey(),
            signedPreKeyPair.getPublicKey().serialize()
        );

        SignedPreKeyRecord signedPreKeyRecord = new SignedPreKeyRecord(
            signedPreKeyId,
            timestamp,
            signedPreKeyPair,
            signedPreKeySignature
        );

        protocolStore.storeSignedPreKey(signedPreKeyId, signedPreKeyRecord);

        // Generate one-time pre-keys
        List<PreKeyRecord> preKeyRecords = KeyHelper.generatePreKeys(PREKEY_START_ID, PREKEY_COUNT);
        for (PreKeyRecord record : preKeyRecords) {
            protocolStore.storePreKey(record.getId(), record);
        }

        Log.i(TAG, "Generated " + PREKEY_COUNT + " pre-keys");
    }

    /**
     * Get pre-key bundle for sharing with other users
     */
    public PreKeyBundle getPreKeyBundle() {
        try {
            IdentityKeyPair identityKeyPair = protocolStore.getIdentityKeyPair();
            int registrationId = protocolStore.getLocalRegistrationId();

            // Get signed pre-key
            SignedPreKeyRecord signedPreKeyRecord = protocolStore.loadSignedPreKey(1);

            // Get a one-time pre-key
            int preKeyId = PREKEY_START_ID + new SecureRandom().nextInt(PREKEY_COUNT);
            PreKeyRecord preKeyRecord = protocolStore.loadPreKey(preKeyId);

            return new PreKeyBundle(
                registrationId,
                1, // device ID
                preKeyId,
                preKeyRecord.getKeyPair().getPublicKey(),
                signedPreKeyRecord.getId(),
                signedPreKeyRecord.getKeyPair().getPublicKey(),
                signedPreKeyRecord.getSignature(),
                identityKeyPair.getPublicKey()
            );
        } catch (Exception e) {
            Log.e(TAG, "Error getting pre-key bundle", e);
            return null;
        }
    }

    /**
     * Process received pre-key bundle from another user
     */
    public void processPreKeyBundle(String recipientId, PreKeyBundle preKeyBundle) {
        try {
            SignalProtocolAddress recipientAddress = new SignalProtocolAddress(recipientId, 1);

            org.whispersystems.libsignal.SessionBuilder sessionBuilder =
                new org.whispersystems.libsignal.SessionBuilder(protocolStore, recipientAddress);

            sessionBuilder.process(preKeyBundle);

            Log.i(TAG, "Processed pre-key bundle for " + recipientId);
        } catch (Exception e) {
            Log.e(TAG, "Error processing pre-key bundle", e);
        }
    }

    /**
     * Encrypt a message for a recipient
     */
    public String encryptMessage(String recipientId, String plaintext) {
        try {
            SignalProtocolAddress recipientAddress = new SignalProtocolAddress(recipientId, 1);
            SessionCipher sessionCipher = new SessionCipher(protocolStore, recipientAddress);

            CiphertextMessage ciphertext = sessionCipher.encrypt(plaintext.getBytes());

            // Convert to base64 for transmission
            byte[] serialized = ciphertext.serialize();
            String encoded = Base64.encodeToString(serialized, Base64.NO_WRAP);

            // Prepend message type (3 = PreKeySignalMessage, 2 = SignalMessage)
            String prefix = ciphertext.getType() == CiphertextMessage.PREKEY_TYPE ? "3:" : "2:";

            Log.i(TAG, "Encrypted message for " + recipientId);
            return prefix + encoded;

        } catch (UntrustedIdentityException e) {
            Log.e(TAG, "Untrusted identity for " + recipientId, e);
            return null;
        }
    }

    /**
     * Decrypt a received message
     */
    public String decryptMessage(String senderId, String ciphertextString) {
        try {
            // Parse message type and ciphertext
            String[] parts = ciphertextString.split(":", 2);
            if (parts.length != 2) {
                Log.e(TAG, "Invalid ciphertext format");
                return null;
            }

            int messageType = Integer.parseInt(parts[0]);
            byte[] ciphertext = Base64.decode(parts[1], Base64.NO_WRAP);

            SignalProtocolAddress senderAddress = new SignalProtocolAddress(senderId, 1);
            SessionCipher sessionCipher = new SessionCipher(protocolStore, senderAddress);

            byte[] plaintext;

            if (messageType == CiphertextMessage.PREKEY_TYPE) {
                // PreKey message (first message in a conversation)
                PreKeySignalMessage preKeyMessage = new PreKeySignalMessage(ciphertext);
                plaintext = sessionCipher.decrypt(preKeyMessage);
            } else {
                // Regular message
                SignalMessage signalMessage = new SignalMessage(ciphertext);
                plaintext = sessionCipher.decrypt(signalMessage);
            }

            Log.i(TAG, "Decrypted message from " + senderId);
            return new String(plaintext);

        } catch (InvalidMessageException | InvalidKeyException | InvalidKeyIdException |
                 DuplicateMessageException | InvalidVersionException | LegacyMessageException |
                 NoSessionException | UntrustedIdentityException e) {
            Log.e(TAG, "Error decrypting message from " + senderId, e);
            return null;
        }
    }

    /**
     * Check if we have a session with a user
     */
    public boolean hasSession(String userId) {
        SignalProtocolAddress address = new SignalProtocolAddress(userId, 1);
        return protocolStore.containsSession(address);
    }

    /**
     * Delete session with a user
     */
    public void deleteSession(String userId) {
        SignalProtocolAddress address = new SignalProtocolAddress(userId, 1);
        protocolStore.deleteSession(address);
        Log.i(TAG, "Deleted session with " + userId);
    }

    /**
     * Get the public identity key as a base64 string
     */
    public String getPublicIdentityKey() {
        try {
            IdentityKeyPair identityKeyPair = protocolStore.getIdentityKeyPair();
            byte[] publicKeyBytes = identityKeyPair.getPublicKey().serialize();
            return Base64.encodeToString(publicKeyBytes, Base64.NO_WRAP);
        } catch (Exception e) {
            Log.e(TAG, "Error getting public identity key", e);
            return null;
        }
    }

    /**
     * Save another user's public identity key
     */
    public void savePublicIdentityKey(String userId, String publicKeyBase64) {
        try {
            byte[] publicKeyBytes = Base64.decode(publicKeyBase64, Base64.NO_WRAP);
            IdentityKey identityKey = new IdentityKey(publicKeyBytes, 0);

            SignalProtocolAddress address = new SignalProtocolAddress(userId, 1);
            protocolStore.saveIdentity(address, identityKey);

            Log.i(TAG, "Saved public identity key for " + userId);
        } catch (Exception e) {
            Log.e(TAG, "Error saving public identity key", e);
        }
    }

    /**
     * Clear all encryption data (for logout)
     */
    public void clearAll() {
        protocolStore.deleteAllSessions();
        Log.i(TAG, "Cleared all encryption data");
    }
}
